File: tests/abi_probe_hisi5.py
#!/usr/bin/env python3
"""Print the sizeof() table a cross-compiled abi_probe_hisi5.c left in .rodata.sizes.

    arm-openipc-linux-musleabi-gcc -std=c11 -I<openhisilicon>/kernel/include/hi3516cv6xx -c -o abi_probe_hisi5.o abi_probe_hisi5.c
    arm-openipc-linux-musleabi-objcopy -O binary --only-section=.rodata.sizes abi_probe_hisi5.o abi_probe_hisi5.bin
    python3 abi_probe_hisi5.py abi_probe_hisi5.bin

The name list must match the order of the sizes[] initialiser in abi_probe_hisi5.c.
"""
import struct, sys

NAMES = ("MAGIC ot_vi_dev_attr ot_vi_pipe_attr ot_vi_chn_attr ot_vi_vpss_mode "
         "ot_vpss_grp_attr ot_vpss_chn_attr ot_venc_chn_attr ot_venc_stream ot_venc_pack "
         "ot_venc_chn_status ot_venc_rc_param ot_vb_cfg ot_vb_pool_cfg ot_mpp_chn "
         "ot_isp_pub_attr ot_isp_exp_info ot_isp_3a_alg_lib ot_isp_sns_obj ot_isp_sns_commbus "
         "ot_rgn_attr ot_rgn_chn_attr ot_rgn_canvas_info ot_aio_attr ot_audio_frame "
         "combo_dev_attr_t ot_mpp_version "
         "ot_size ot_rect ot_vb_pool_info ot_vb_pool_status ot_vb_supplement_cfg "
         "ot_video_frame ot_video_frame_info ot_video_supplement ot_frame_rate_ctrl "
         "ot_border ot_aspect_ratio mipi_dev_attr_t img_rect_t ot_vi_sync_cfg "
         "ot_vi_timing_blank ot_vpss_crop_info ot_vpss_grp_param ot_venc_attr "
         "ot_venc_rc_attr ot_venc_mjpeg_fixqp ot_venc_gop_attr ot_venc_pack_info ot_venc_start_param "
         "ot_venc_chn_param ot_venc_jpeg_attr ot_venc_jpeg_param ot_venc_mpf_cfg "
         "ot_venc_stream_buf_info ot_mipi_crop_attr ot_isp_sns_attr_info "
         "ot_crop_info").split()

b = open(sys.argv[1], "rb").read()
v = struct.unpack("<%dI" % (len(b) // 4), b)
if v[0] != 0xC0DE0001:
    sys.exit("bad magic %#x: wrong section or wrong endianness" % v[0])
if len(v) != len(NAMES):
    sys.exit("%d values for %d names: abi_probe_hisi5.c and this list disagree" % (len(v), len(NAMES)))
for n, x in zip(NAMES[1:], v[1:]):
    print("%-22s %6d" % (n, x))

File: tests/abi_probe_hisi5.c
/*
 * abi_probe_hisi5.c -- the size oracle for the HiMPP V5 transcriptions.
 *
 * src/hisi_v5/v5_*.h declares the V5 ABI itself, the way src/hisi_v4/v4_*.h
 * does, and pins every layout with _Static_assert. This file is where those
 * numbers come from: it is compiled with the ARM cross-compiler against the
 * vendor headers -- openhisilicon's GPL copy of the 1.0.2.0 set, which is the
 * version OpenIPC's hi3516cv6xx libraries were built from -- and leaves a
 * table of sizeof() in its own section for abi_probe_hisi5.py to read back.
 * Nothing is linked and nothing runs on a board.
 *
 * It is a *probe*, not a check: it says what the headers claim, and the
 * headers can disagree with the libraries. abi-check-hisi5 (Phase 1) is the
 * check, and it compiles the transcriptions rather than the vendor headers.
 *
 * The sizes[] order is the contract with abi_probe_hisi5.py's name list; the
 * magic and the count are what catch the two of them drifting apart.
 *
 * SPDX-License-Identifier: MIT
 */
#include "ot_common_vi.h"
#include "ot_common_vpss.h"
#include "ot_common_venc.h"
#include "ot_common_vb.h"
#include "ot_common_sys.h"
#include "ot_common_isp.h"
#include "ot_common_3a.h"
#include "ot_sns_ctrl.h"
#include "ot_common_region.h"
#include "ot_common_aio.h"
#include "ot_mipi_rx.h"
#define S(t) sizeof(t)
const unsigned int sizes[] __attribute__((section(".rodata.sizes"))) = {
    0xC0DE0001,
    S(ot_vi_dev_attr),
    S(ot_vi_pipe_attr),
    S(ot_vi_chn_attr),
    S(ot_vi_vpss_mode),
    S(ot_vpss_grp_attr),
    S(ot_vpss_chn_attr),
    S(ot_venc_chn_attr),
    S(ot_venc_stream),
    S(ot_venc_pack),
    S(ot_venc_chn_status),
    S(ot_venc_rc_param),
    S(ot_vb_cfg),
    S(ot_vb_pool_cfg),
    S(ot_mpp_chn),
    S(ot_isp_pub_attr),
    S(ot_isp_exp_info),
    S(ot_isp_3a_alg_lib),
    S(ot_isp_sns_obj),
    S(ot_isp_sns_commbus),
    S(ot_rgn_attr),
    S(ot_rgn_chn_attr),
    S(ot_rgn_canvas_info),
    S(ot_aio_attr),
    S(ot_audio_frame),
    S(combo_dev_attr_t),
    S(ot_mpp_version),
    /* Appended by Phase 1, which transcribed them: the two geometry structs
     * every attribute embeds, and the VB readback trio v5_vb.h declares.
     * Appended rather than inserted -- the order is the contract with
     * abi_probe_hisi5.py's name list, so a new struct goes at the end. */
    S(ot_size),
    S(ot_rect),
    S(ot_vb_pool_info),
    S(ot_vb_pool_status),
    S(ot_vb_supplement_cfg),
    /* Appended by Phase 2, in the order v5_video.h, v5_mipi.h, v5_vi.h,
     * v5_vpss.h, v5_venc.h, v5_isp.h transcribe them. Same rule as above:
     * the order is the contract, so new entries go at the end. */
    S(ot_video_frame),
    S(ot_video_frame_info),
    S(ot_video_supplement),
    S(ot_frame_rate_ctrl),
    S(ot_border),
    S(ot_aspect_ratio),
    S(mipi_dev_attr_t),
    S(img_rect_t),
    S(ot_vi_sync_cfg),
    S(ot_vi_timing_blank),
    S(ot_vpss_crop_info),
    S(ot_vpss_grp_param),
    S(ot_venc_attr),
    S(ot_venc_rc_attr),
    S(ot_venc_gop_attr),
    S(ot_venc_pack_info),
    S(ot_venc_start_param),
    S(ot_venc_chn_param),
    S(ot_venc_jpeg_attr),
    S(ot_venc_jpeg_param),
    S(ot_venc_mpf_cfg),
    S(ot_venc_stream_buf_info),
    S(ot_mipi_crop_attr),
    S(ot_isp_sns_attr_info),
    S(ot_crop_info),
};
